transaction-payment/src/lib.rs

```diff
@@ -0,0 +1,285 @@
+    /// Utility constructor. Used only in client/factory code.
+    pub fn from(tip: BalanceOf<T>, asset_id: Option<ChargeAssetIdOf<T>>) -> Self {
+        Self { tip, asset_id }
+    }
+
+    /// Fee withdrawal logic that dispatches to either `OnChargeAssetTransaction` or
+    /// `OnChargeTransaction`.
+    fn withdraw_fee(
+        &self,
+        who: &T::AccountId,
+        call: &T::RuntimeCall,
+        info: &DispatchInfoOf<T::RuntimeCall>,
+        len: usize,
+    ) -> Result<(BalanceOf<T>, InitialPayment<T>), TransactionValidityError> {
+        let fee = pallet_transaction_payment::Pallet::<T>::compute_fee(len as u32, info, self.tip);
+        debug_assert!(self.tip <= fee, "tip should be included in the computed fee");
+        if fee.is_zero() {
+            Ok((fee, InitialPayment::Nothing))
+        } else if let Some(asset_id) = self.asset_id {
+            <T as pallet::Config>::OnChargeAssetTransaction::withdraw_fee(
+                who,
+                call,
+                info,
+                asset_id,
+                fee.into(),
+                self.tip.into(),
+            )
+            .map(|i| (fee, InitialPayment::Asset(i.into())))
+        } else {
+            <OnChargeTransactionOf<T> as OnChargeTransaction<T>>::withdraw_fee(
+                who, call, info, fee, self.tip,
+            )
+            .map(|i| (fee, InitialPayment::Native(i)))
+            .map_err(|_| -> TransactionValidityError { InvalidTransaction::Payment.into() })
+        }
+    }
+}
+
+impl<T: Config> sp_std::fmt::Debug for ChargeAssetTxPayment<T> {
+    #[cfg(feature = "std")]
+    fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+        write!(f, "ChargeAssetTxPayment<{:?}, {:?}>", self.tip, self.asset_id.encode())
+    }
+    #[cfg(not(feature = "std"))]
+    fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+        Ok(())
+    }
+}
+
+impl<T: Config> SignedExtension for ChargeAssetTxPayment<T>
+where
+    T::RuntimeCall: Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo>
+        + IsSubType<pallet_ismp::Call<T>>,
+    T::Hash: From<H256>,
+    AssetBalanceOf<T>: Send + Sync + FixedPointOperand,
+    BalanceOf<T>: Send + Sync + From<u64> + FixedPointOperand + IsType<ChargeAssetBalanceOf<T>>,
+    ChargeAssetIdOf<T>: Send + Sync,
+    Credit<T::AccountId, <T as pallet_asset_tx_payment::Config>::Fungibles>:
+        IsType<ChargeAssetLiquidityOf<T>>,
+{
+    const IDENTIFIER: &'static str = "ChargeAssetTxPayment";
+    type AccountId = T::AccountId;
+    type Call = T::RuntimeCall;
+    type AdditionalSigned = ();
+    type Pre = (
+        // tip
+        BalanceOf<T>,
+        // who paid the fee
+        Self::AccountId,
+        // imbalance resulting from withdrawing the fee
+        InitialPayment<T>,
+        // asset_id for the transaction payment
+        Option<ChargeAssetIdOf<T>>,
+        // boolean to indicate whether the call is an ISMP call
+        IsIsmpCall,
```

Comment:
Return an Option<Self> instead, if it's defined in then we'll withdraw fee in `post_dispatch`.